com.unity.shadergraph/Editor/ShaderFoundry/Linker/LegacyTemplateLinker.cs

```diff
@@ -77,22 +96,44 @@ void GenerateSubShaderTags(UnityEditor.ShaderGraph.SubShaderDescriptor descripto
             }
         }
 
-        void GenerateShaderPass(Template template, TemplatePass pass, IEnumerable<CustomizationPointInstance> customizationPointInstances, ShaderBuilder builder)
+        LegacyEntryPoints GeneratePassBlocks(Template template, TemplatePass pass, IEnumerable<CustomizationPointInstance> customizationPointInstances)
         {
-            UnityEditor.ShaderGraph.PassDescriptor legacyPass = new UnityEditor.ShaderGraph.PassDescriptor();
-            if (!FindLegacyPass(pass.ReferenceName, ref legacyPass))
-                throw new Exception("Shouldn't happen");
-
             var passCustomizationPointInstances = FindCustomizationPointsForPass(pass, customizationPointInstances);
 
             var legacyBlockLinker = new SimpleLegacyBlockLinker(Container);
             var legacyEntryPoints = legacyBlockLinker.GenerateLegacyEntryPoints(template, pass, passCustomizationPointInstances);
+            return legacyEntryPoints;
+        }
+
+        void BuildShaderProperties(IEnumerable<LegacyEntryPoints> legacyEntryPoints, ShaderPropertyCollection shaderProperties)
+        {
+            void AddBlockInstanceProperties(BlockInstance blockInstance)
+            {
+                if (!blockInstance.IsValid)
+                    return;
+
+                foreach (var property in blockInstance.Block.Properties())
+                    shaderProperties.Add(property);
+            }
+
+            foreach (var legacyEntryPoint in legacyEntryPoints)
+            {
+                AddBlockInstanceProperties(legacyEntryPoint.vertexDescBlockInstance);
+                AddBlockInstanceProperties(legacyEntryPoint.fragmentDescBlockInstance);
+            }
+        }
+
+        void GenerateShaderPass(Template template, TemplatePass pass, LegacyEntryPoints legacyEntryPoints, ShaderPropertyCollection shaderProperties, ShaderBuilder builder)
+        {
+            UnityEditor.ShaderGraph.PassDescriptor legacyPass = new UnityEditor.ShaderGraph.PassDescriptor();
+            if (!FindLegacyPass(pass.ReferenceName, ref legacyPass))
+                throw new Exception("Shouldn't happen");
```

Comment:
This could be a bit more descriptive (e.g., `$"Unable to find pass {pass.ReferenceName}."`)

Reply:
That's actually an old piece of code that I just moved. I'll update it though to be more descriptive